raft/src/test/java/org/apache/kafka/raft/RaftEventSimulationTest.java

```diff
@@ -1127,14 +1331,75 @@ private MajorityReachedHighWatermark(Cluster cluster) {
 
         @Override
         public void verify() {
-            cluster.leaderHighWatermark().ifPresent(highWatermark -> {
-                long numReachedHighWatermark = cluster.nodes.entrySet().stream()
-                    .filter(entry -> cluster.voters.containsKey(entry.getKey()))
-                    .filter(entry -> entry.getValue().log.endOffset().offset() >= highWatermark)
-                    .count();
-                assertTrue(
-                    numReachedHighWatermark >= cluster.majoritySize(),
-                    "Insufficient nodes have reached current high watermark");
+            if (cluster.withKip853) {
+                /*
+                * For clusters running in KIP-853 mode, we check that a majority of at least one of:
+                * 1. the leader's voter set at the HWM
```

Comment:
nit: wording is a bit unclear (e.g. reading "check that a majority of at least one of the leader's voter set at the HWM" doesn't make sense to me)

Reply:
I see, the sentence continues on L1339

wonder if something like the following might be more clear? maybe you can fix my language a bit too

```suggestion
                * For clusters running in KIP-853 mode, we consider two possibilities for what the majority of the cluster is:
                * 1. the majority based off the current value of lastVoterSet()
                * 2. the majority based off what the leader's voter set was at the HWM (i.e. the prior lastVoterSet() value)
                * This is because verify() may be called after the leader's in-memory voter set was updated but before the voter change was committed. A perfect invariant checker would use what has been committed to the log but that information is not readily exposed for testing.
```

Reply:
Sorry, upon further looking into this, I think the issue is just slightly different that what I've described thus far.

Essentially, when we fail the invariant check when only using `lastVoterSet()`, what's going on is that the caller of `verify()` is a different thread (lets say event scheduler thread) than the KRaft leader's raft-io thread which is continuously doing all of this internal state updating. The event scheduler thread is looking at a bunch of the leader's internal state (e.g. partitionState and highWatermark), which can definitely be in the following state: `partitionState` has been updated with a new voter set, but a new `highWatermark` value has not yet been calculated with this voter set yet, which could cause this invariant check to fail when only looking at `lastVoterSet()`. The leader may not be in this state for very long, since it's in this state in between the calls to `appendAsLeader -> updateState -> maybeLoadLog` and `flushLeaderLog -> maybeUpdateHighWatermark`, but it nevertheless looks like a valid state to me.

Invariants as a concept are predicates that should be true for the system at all times, no matter when you check them, so I think performing invariant verification how it's currently implemented is fine, as we could be checking the internal states of our raft nodes at any point in their execution. However, it just means we have to consider these "intermediary" states when performing verification.